### PR TITLE
feat: like prediate and like udf

### LINKED ISSUE #224: sql: support like predicate
**Is your feature request related to a problem? Please describe.**

Currently, OpenMLDB doesn't support the some basic SQL syntaxes like `HAVING` clause and `LIKE` operator. There abilities are required by data scientis for data analysis.
> A **`HAVING`** clause in [SQL](https://en.wikipedia.org/wiki/SQL) specifies that an SQL `SELECT` statement must only return rows where aggregate values meet the specified conditions.

Generally, the `HAVING` clause is used in conjunction with aggregate functions and the GROUP BY clause to control which groups are selected.

## Support `LIKE` operator

The LIKE operator supports these wildcard characters in the quoted string.

#### Syntax

```
[ NOT ] { LIKE search_pattern [ ESCAPE esc_char ] }
```

#### Parameters

- **search_pattern**

  Specifies a string pattern to be searched by the `LIKE` clause. It can contain special pattern-matching characters:

  - `%` matches zero or more characters.
  - `_` matches exactly one character.

- **esc_char**

  Specifies the escape character. The default escape character is `\`.

- **regex_pattern**

  Specifies a regular expression search pattern to be searched by the `RLIKE` clause.

### Examples:

```SQL
CREATE TABLE person (id INT, name STRING, age INT);
INSERT INTO person VALUES
    (100, 'John', 30),
    (200, 'Mary', NULL),
    (300, 'Mike', 80),
    (400, 'Dan',  50),
    (500, 'Evan_w', 16);

SELECT * FROM person WHERE name LIKE 'M%';
+---+----+----+
| id|name| age|
+---+----+----+
|300|Mike|  80|
|200|Mary|null|
+---+----+----+

SELECT * FROM person WHERE name LIKE 'M_ry';
+---+----+----+
| id|name| age|
+---+----+----+
|200|Mary|null|
+---+----+----+

SELECT * FROM person WHERE name NOT LIKE 'M_ry';
+---+------+---+
| id|  name|age|
+---+------+---+
|500|Evan_W| 16|
|300|  Mike| 80|
|100|  John| 30|
|400|   Dan| 50|
+---+------+---+


SELECT * FROM person WHERE name LIKE '%\_%';
+---+------+---+
| id|  name|age|
+---+------+---+
|500|Evan_W| 16|
+---+------+---+

SELECT * FROM person WHERE name LIKE '%$_%' ESCAPE '$';
+---+------+---+
| id|  name|age|
+---+------+---+
|500|Evan_W| 16|
+---+------+---+
```

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### LINKED ISSUE #769: Support new SQL syntax for job management
Since we want to support job management in OpenMLDB CLI. The SQL parser needs to be extends and add the following commands.

```
SHOW JOBS

SHOW JOB $id

STOP JOB $id

DELETE JOB $id
```

More detailed in https://github.com/4paradigm/OpenMLDB/discussions/732 .

### LINKED ISSUE #799: Add database name for PhysicalDataProviderNode
Now we need to know the database name when resolving SQL and get PhysicalDataProviderNode. It would be better to add the database name for PhysicalDataProviderNode.

```
class PhysicalDataProviderNode : public PhysicalOpNode {
 public:
    PhysicalDataProviderNode(const std::shared_ptr<TableHandler> &table_handler,
                             DataProviderType provider_type)
        : PhysicalOpNode(kPhysicalOpDataProvider, true),
          provider_type_(provider_type),
          table_handler_(table_handler) {}
    ~PhysicalDataProviderNode() {}

    base::Status InitSchema(PhysicalPlanContext *) override;

    static PhysicalDataProviderNode *CastFrom(PhysicalOpNode *node);
    const std::string &GetName() const;
    const DataProviderType provider_type_;
    const std::shared_ptr<TableHandler> table_handler_;
};
```

### PR DESCRIPTION
**Pull request by aceforeverd**
### Features

resolve #224 
detail sql rules found in #224 and #686 

### What not implemented yet:

- convert <match character> and <pattern> into string if possible, currently only support string or null
- data exception: invalid escape sequence is not checked

### Further work

- Implicit conversion for like's `<target expression>` & `<pattern expression>`
- support `RLIKE`/`SIMILAR TO` predicate (regexp_match)

### CODE REVIEW

File: hybridse/src/udf/udf.cc
Comment (aceforeverd) [inline hybridse/src/udf/udf.cc]: not escape

File: hybridse/src/udf/udf.cc
Comment (jingchen2222) [inline hybridse/src/udf/udf.cc]: The rules are quite different from what issue #224 described. I think that is because #224 borrows rules from T-SQL but the current implemented rules follow MySQL's rules to some extend.
- A little concern about the rule that the pattern is default case insensitive. What if the users try to match a string starting with 'A' exactly? I have noticed the `regexp_match ` function but didn't find how to invoke it?
- In my opinion, T-SQL's rules are much easier to implement.
   - pattern is case sensitive by default.
   - using LIKE '[Aa]%' to match a string starting with 'a' or 'A'
   - using LIKE 'A%' to match a string starting with 'A' exactly.

However, since there is no standard for the LIKE pattern for SQL, so both kinds of rules are ok. 

Please update #224 with our new rules to keep consistency. 
Update related SQL documentation after the pr is merged.

File: hybridse/src/udf/udf_test.cc
Comment (jingchen2222) [inline hybridse/src/udf/udf_test.cc:667]: Can we perform `LIKE` on Chinese characters or other Unicode chars?
If so, add some related unit tests and YAML cases.
If non-support currently, add some comments and documentation. And throw an error during converting ASTNode or expression evaluation.

File: hybridse/src/udf/default_udf_library.cc
Comment (jingchen2222) [inline hybridse/src/udf/default_udf_library.cc]: Please add enough rules and examples for "like_match" in `.doc()`, since it is really important and there are no standard rules for that.

File: hybridse/src/udf/udf.cc
Comment (aceforeverd) [reply hybridse/src/udf/udf.cc]: I follow the ansi sql and postgre sql's guide as well: https://www.postgresql.org/docs/current/functions-matching.html

That detail of like predicate may different in many ways, but should easy to tweak, e.g for case sensitive just change the equal function.

bracket character is not implemented, it seems not the standard of ansi sql.

Comment (aceforeverd) [reply hybridse/src/udf/udf.cc]: BTW, there does has a general problem:

It may not appropriate to use backslash(`\`) as default escape character, since zetasql parser backslash as escape char as well. As a result, in string `'abc\n'`, `'\n'` is treated as special character, and in a like predicate, we must write `col1 LIKE 'abc\\n' ESCAPE '\\` to represent that backslash. Also, in yaml, `"abc\\n"` seems have different meanings.

So I'd wondering if we should tweak a lit of escape default:
1. if `ESCAPE` clause not exist, disable escape feature
2. documented: in sql, use `\\` to represent backslash
3. There is no default escape character and users are encourage to escape with another character(prefer not backslash):
    `c1 LIKE 'a$%' ESCAPE '$'`

File: hybridse/src/udf/udf_test.cc
Comment (aceforeverd) [reply hybridse/src/udf/udf_test.cc:667]: I'm afraid not, see #808 , the string going to compare don't have locale info, and the simple u8 char compare is performed

Comment (jingchen2222) [reply hybridse/src/udf/udf_test.cc:667]: It's okay. Inform users in function documentation and our language guide.

**Comment by aceforeverd** [comment]
[sql support for LIKE](https://trello.com/c/ZgAcOGCw/26-sql-support-for-like)

File: hybridse/src/udf/udf.cc
Comment (jingchen2222) [reply hybridse/src/udf/udf.cc]: 1. if `ESCAPE` clause not exist, using '\\' as default escape as described in https://github.com/4paradigm/OpenMLDB/discussions/686.

File: hybridse/src/udf/udf.cc
Comment (aceforeverd) [inline hybridse/src/udf/udf.cc:290]: add a case

File: hybridse/src/udf/udf.cc
Comment (aceforeverd) [inline hybridse/src/udf/udf.cc]: rm unused code

File: hybridse/src/udf/default_udf_library.cc
Comment (aceforeverd) [inline hybridse/src/udf/default_udf_library.cc]: TODO: run with postgre

'Mi_e', 'Mi\_e', '\' -> true  
'Mi_e', 'Mi\_e', '' -> false

File: hybridse/src/udf/udf.cc
Comment (aceforeverd) [inline hybridse/src/udf/udf.cc]: fix this

File: hybridse/src/udf/udf_test.cc
Comment (zhanghaohit) [inline hybridse/src/udf/udf_test.cc]: escape

File: hybridse/src/udf/udf_test.cc
Comment (aceforeverd) [inline hybridse/src/udf/udf_test.cc:721]: add test for meaning of % and _ when escape char disable

File: hybridse/src/udf/udf_test.cc
Comment (aceforeverd) [inline hybridse/src/udf/udf_test.cc:801]: ignore case of escape character 'p' ?

Comment (aceforeverd) [reply hybridse/src/udf/udf_test.cc:801]: no. add a case

File: cases/query/simple_query.yaml
Comment (aceforeverd) [inline cases/query/simple_query.yaml]: string(null)

File: cases/query/simple_query.yaml
Comment (aceforeverd) [inline cases/query/simple_query.yaml:560]: add a case if pattern is columnref

File: hybridse/include/vm/physical_op.h
Comment (aceforeverd) [inline hybridse/include/vm/physical_op.h:565]: add a case

File: cases/query/simple_query.yaml
Comment (jingchen2222) [reply cases/query/simple_query.yaml]: There might be a crush for `string(null)`. https://github.com/4paradigm/OpenMLDB/issues/854.

**Comment by jingchen2222** [comment]
>>> 
> Implicit conversion for like's <target expression> & <pattern expression>
support RLIKE/SIMILAR TO predicate (regexp_match)

Create related issues and link to them.

File: cases/query/simple_query.yaml
Comment (jingchen2222) [reply cases/query/simple_query.yaml]: Will support when #855 is merged

**Review by jingchen2222 — APPROVED** [review]
Add like string(null) cases and Merge when  #855  has been merged

**Review by jingchen2222 — DISMISSED** [review]
Add like string(null) cases when #855 has been merged

**Comment by aceforeverd** [comment]
`string(NULL)` crash, depends #855

**Comment by aceforeverd** [comment]
> Add like string(null) cases when #855 has been merged

BTW, I'd like update behavior of `(i)like_match(___, string(null))`, it default to return `NULL`, instead treat like match as escape mechanism disabled. Which also consistent with ansi sql

**Comment by aceforeverd** [comment]
@jingchen2222  looks like `select like_match(__, string(NULL)) ` still failed, but only on linux( a address boundry error)

**Comment by aceforeverd** [comment]
@jingchen2222  looks like `select like_match(__, string(NULL)) ` still failed, but only on linux( a address boundry error)

**Comment by jingchen2222** [comment]
> @jingchen2222 looks like `select like_match(__, string(NULL)) ` still failed, but only on linux( a address boundry error)
We need to figure out the root cause.  
BTW, `string(NULL)` seems to work on Linux since #855 passed the cicd.

**Comment by aceforeverd** [comment]
> > @jingchen2222 looks like `select like_match(__, string(NULL)) ` still failed, but only on linux( a address boundry error)
> > We need to figure out the root cause.
> > BTW, `string(NULL)` seems to work on Linux since #855 passed the cicd.

Yes,  disable that case temporally to pass cicd. I'll check this problem later

**Comment by aceforeverd** [comment]
@jingchen2222 @dl239 @tobegit3hub this is pr is ready to merge after cicd passed

**Review by tobegit3hub — APPROVED** [review]
LGTM

**Review by dl239 — APPROVED** [review]

**Pull request merged**
